### ISSUE TITLE
Consistent use of (psalm) phpdoc array annotation syntax

### ISSUE THREAD

**Issue by simonberger** [doctrine/doctrine2#9025]
My last state of information was that only a notation  like `string[]` should be used for annotations like `param` and `return`.
I was about to request a change in #9024 but then noticed that those few lines of changed lines revealed three different types of array annotations in the old code:

- `array<string>`
- `array<string, string>`
- `string[]`

It would be great  to have a rule which array syntax is allowed for standard phpdoc annotation and which needs to be put into the psalm variants for more detailed description.
Personally I would prefer to just allow `array<string, string>`- kind syntax everywhere but then we should move them all out of the psalm variants. Otherwise some sort of CI check would be helpful to get some consistency.

As the first and third syntax of the example are the same it would be great to allow just one but that's maybe too pedantic.

**Comment by phansys** [comment]
IMO, the syntax `array<type, type>` is useful when we must tell the type used in the keys. For the rest of the cases, I think the `type[]` syntax is simpler.

**Comment by simonberger** [comment]
@phansys `array<type, type>` has more detail this is not really questionable. The issue is about what is allowed to be used with standard PHPdoc annotations and when psalm-annotations have to be used.

**Comment by greg0ire** [comment]
IMO it should be `array<type, type>` or `list<type>` if the type inside the list doesn't matter. It's kinda unfriendly to consumers that pass the result of `array_filter`, but using `list` shows that we already looked into the type of the keys, and that it shouldn't matter. IMO, any time we would be using a list in another language that supports them (such as Python), we should use `list` here. So I'd say, let's avoid `type[]` entirely.

**Comment by stof** [comment]
`list<type>` does not mean that you don't care about the type of keys. It means you actually want a list, i.e. sequencial integer keys from `0` to `$count - 1` (see `array_is_list` in PHP 8.1). Psalm actually enforces that meaning for `list` (phpstan just treats it as an alias of `array<int, type>` for now)

**Comment by greg0ire** [comment]
@stof yeah but conceptually, it's called a list because that's exactly the data type that is meant (not required, meant) to be passed here. If you use the PHP syntax to create a list, you will end up with an array that happens to keys in a sequence.

So while on the one hand, using lists add an unwanted requirement, on the other hand, it lets the user know that they should not expect anything special to happen if they pass a map instead, and might even let them know they have a bug in their code because of that expectation.

**Comment by stof** [comment]
@greg0ire if we really don't care about the keys (some places do care), I'd rather use `type[]` than `list<type>`
